### ISSUE TITLE
Improve docstring of EpochMetric

### ISSUE THREAD

**Issue by vfdev-5** [pytorch/ignite#309]
As the current implementation stores all predictions and targets in RAM we need to explicitly detail such behaviour and note that this could lead to a memory error depending on the size of the input data.

**Comment by vfdev-5** [comment]
Fixed by #314

**Comment by ZhiliangWu** [comment]
Hi @vfdev-5, I used `EpochMetric()` as a utility to quickly save the entire output history so as they can be used for possible plotting functions (maybe the use case is a bit rare). Therefore, I defined the `compute_fn` as `lambda y_pred, y_true: torch.stack([y_pred, y_true], dim=1)`, where `y_pred.shape=(#samples, )`.  Things seem to work as expected. 

My point here is, maybe the [documentation for `EpochMetric()`](https://pytorch.org/ignite/metrics.html#ignite.metrics.EpochMetric)

> compute_fn (callable): a callable with the signature (`torch.tensor`, `torch.tensor`) takes as the input `predictions` and `targets` and **returns a scalar**. 

doesn't have to be restricted to output a `scalar`, though it's the most obvious use case. Thanks for your comments in advance.

**Comment by vfdev-5** [comment]
Hi Zhiliang,

I think EpochMetric should remain a metric and compute a scalar or vector.
Maybe it would make more  sense to introduce some specific handler to store
all output history. What do you think ?

**Comment by ZhiliangWu** [comment]
Hi @vfdev-5, yes, it would make more sense to have something like `entire_history` handler. Actually it took me a while to finally realize the `EpochMetric()` is able to do that, though it's designed for some more advanced computation usage.

**Comment by vfdev-5** [comment]
Hi Zhiliang, if you would like to contribute a class, e.g.
ignite.handlers.stores.EpochOutputStore that appends engine's output into a
list during one epoch, it could help the library development :)

**Comment by ZhiliangWu** [comment]
Hi @vfdev-5, I am happy to contribute if possible. Currently, I am a bit confused about whether it should be a handler or a metric. Maybe you can help me with it.
**From the implementation point of view it looks like a metric**. As the output values are accumulated in an online fashion from the `process_function` like any other metrics. It would be a subclass inherited from `EpochMetric()` and overwrites the `compute_fn` with some concatenation operation, which is quite straightforward to write. 
**From the definition, it's better to be a separate handler to save the output into a list after each epoch.** However, the input to handlers is usually `Engine` like `trainer`. I.e., as a separate handler, it has to include something like `evaluator(train_dataloader)` again, which involves one more unnecessary pass-through the data loaders if there already exists handlers to compute metrics. 
Do you have any good ideas about it?

**Comment by vfdev-5** [comment]
Hi Zhiliang, thanks for pointing out those points !
Probably, we need to review again the use-case where we could need a
structure like `EpochMetric` to store the data. What is exactly your
use-case where you use `EpochMetric` as utility without computing a metric
value. The drawbacks of this approach is extensive memory consumption. In
your use-case, can you perform plotting per iteration instead of epoch-wise
?

> *From the implementation point of view it looks like a metric*

Yes, internal implementation can look like `Metric` and has an `attach`
method. But, for instance it is not a problem.

> *From the definition, it's better to be a separate handler to save the
output into a list after each epoch.*

With "attach"-like usage we can store the data within a single pass. I'm
not sure that we should pass twice over the data.
The usage I thought is something like that
```python
trainer = ...

store = EpochOutputStore()
# store.data is empty list : []
store.attach(trainer)

trainer.run(train_dataloader)

# now store.data contains [output1, output2, ...]
```

What do you think ?

**Comment by ZhiliangWu** [comment]
Hi @vfdev-5, actually yes. I have to admit it's quick and dirty usage, where I wanted to investigate some regression metrics visually with [prediction error plot API provided by yellowbrick](https://www.scikit-yb.org/en/latest/api/regressor/peplot.html). As everything is based on matplotlib, incrementally plotting should be possible. Though online updating is more elegant, I feel it's okay to use all history for quick investigations as most visualization packages accept all data points at once.

**Comment by vfdev-5** [comment]
Hi @ZhiliangWu yes, the use-case makes sense and we can provide some simplification handlers like described above. Anyway, let's implement it as a handler with three methods: reset, update and attach (where we add reset and update to provided engine). This can be the first iteration and after we can refactor the code with appropriate interfaces.
